### PR TITLE
Improve Decimal to Double cast performance

### PR DESCRIPTION
**Pull request by majetideepak**
Apply Decimal to Double cast at the batch level instead of row-level.
This also improves performance of cast between primitive types.

### CODE REVIEW

File: velox/expression/CastExpr.cpp [hunk 1]
```diff
@@ -134,6 +123,29 @@ void applyIntToDecimalCastKernel(
...
+  VectorPtr result;
+  context.ensureWritable(rows, DOUBLE(), result);
+  (*result).clearNulls(rows);
+  auto sourceVector = input.as<SimpleVector<TInput>>();
+  auto resultBuffer =
+      result->asUnchecked<FlatVector<double>>()->mutableRawValues();
+  auto toPrecisionScale = getDecimalPrecisionScale(*fromType);
+  context.applyToSelectedNoThrow(rows, [&](int row) {
+    auto output =
+        util::Converter<CppToType<double>::typeKind, void, false>::cast(
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:141]: Let's avoid CppToType.

CppToType<double>::typeKind -> TypeKind::DOUBLE

File: velox/expression/CastExpr.cpp [hunk 2]
```diff
@@ -134,6 +123,29 @@ void applyIntToDecimalCastKernel(
...
+            input.as<SimpleVector<TInput>>()->valueAt(row));
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:142]: Let's move this dynamic cast outside of the loop.

```
auto simpleInput = input.as<SimpleVector<TInput>>();
context.applyToSelectedNoThrow...
```

File: velox/expression/CastExpr.cpp [hunk 3]
```diff
@@ -134,6 +123,29 @@ void applyIntToDecimalCastKernel(
...
+    resultBuffer[row] =
+        output / (double)DecimalUtil::kPowersOfTen[toPrecisionScale.second];
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:144]: Let's move (double)DecimalUtil::kPowersOfTen[toPrecisionScale.second] outside the loop as well.

**Review by mbasmanova — APPROVED** [review]
@majetideepak Thanks. It would be nice to add a benchmark in a follow-up PR.

CC: @laithsakka @kevinwilfong

File: velox/expression/CastExpr.cpp [hunk 4]
```diff
@@ -135,13 +135,13 @@ VectorPtr applyDecimalToDoubleCast(
   auto sourceVector = input.as<SimpleVector<TInput>>();
   auto resultBuffer =
       result->asUnchecked<FlatVector<double>>()->mutableRawValues();
-  auto toPrecisionScale = getDecimalPrecisionScale(*fromType);
+  auto [precision, scale] = getDecimalPrecisionScale(*fromType);
+  const auto simpleInput = input.as<SimpleVector<TInput>>();
+  const auto kDenominator = (double)DecimalUtil::kPowersOfTen[scale];
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:140]: naming: kXxx is for static constants

kDenominator -> denominator

Comment (majetideepak) [reply velox/expression/CastExpr.cpp]: my bad. Thanks. Fixed.

**Comment by majetideepak** [comment]
@mbasmanova I noticed `(double)DecimalUtil::kPowersOfTen[scale];` is not required since `output` is of `double` type.
I fixed this. Can you import again? Thanks.

**Comment by karteekmurthys** [comment]
@majetideepak Thanks for adding this. Was planning on taking up this change. I had mentioned here: https://github.com/facebookincubator/velox/pull/4594

**Comment by mbasmanova** [comment]
@majetideepak Curious, did we not get a signal from conbench on the original PR that introduced the regression?

CC: @pedroerp @laithsakka @kevinwilfong

**Comment by majetideepak** [comment]
@mbasmanova We did after merge. Looks like we should add an option to run conbench before merge.

**Comment by pedroerp** [comment]
@mbasmanova @majetideepak conbench does run the same jobs at PR time (and flagged the same regression) but it's currently not configured to fail the job. Presumably this was done while we were working on stabilizing the benchmark results to prevent false positions. 

I'll sync up with @kgpai and @assignUser when they are back from PTO to enable this. @mbasmanova also had a good point that the results we get in the PR are a bit crypt for now, so I'll chat with them to try to provide something that's easier to digest.

**Pull request closed without merging**